File: rust/crates/symphony-core/src/issue.rs
//! SPEC §4.1.1 normalized issue record.

use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

/// Blocker reference (SPEC §4.1.1 `blocked_by`).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
pub struct Blocker {
    pub id: Option<String>,
    pub identifier: Option<String>,
    pub state: Option<String>,
}

/// Normalized issue record used by orchestration, prompt rendering, and
/// observability output.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Issue {
    pub id: String,
    pub identifier: String,
    pub title: String,
    pub description: Option<String>,
    pub priority: Option<i32>,
    pub state: String,
    pub branch_name: Option<String>,
    pub url: Option<String>,
    /// Always lowercase per SPEC §11.3.
    pub labels: Vec<String>,
    pub blocked_by: Vec<Blocker>,
    #[serde(with = "time::serde::rfc3339::option")]
    pub created_at: Option<OffsetDateTime>,
    #[serde(with = "time::serde::rfc3339::option")]
    pub updated_at: Option<OffsetDateTime>,
}

impl Issue {
    /// Lower-cased view of `state` for SPEC §4.2 normalized state comparisons.
    pub fn normalized_state(&self) -> String {
        self.state.to_lowercase()
    }

    /// Whether all entries in `blocked_by` reference issues in a terminal
    /// state, taking the configured `terminal_states` (SPEC §8.2 blocker rule).
    pub fn blockers_all_terminal(&self, terminal_states: &[String]) -> bool {
        self.blocked_by.iter().all(|b| match &b.state {
            None => false,
            Some(s) => terminal_states.iter().any(|t| t.eq_ignore_ascii_case(s)),
        })
    }
}

File: go/internal/issue/issue.go
// Package issue defines the canonical Issue and Blocker types per SPEC §4.1.1.
package issue

import (
	"strings"
	"time"
)

// Issue is the orchestrator's view of one tracked work item.
//
// All fields except ID, Identifier, Title, and State are optional; the
// dispatch eligibility predicate (§16.2) treats missing required fields
// as a hard fail rather than coercing them.
type Issue struct {
	ID          string     `json:"id"`
	Identifier  string     `json:"identifier"`
	Title       string     `json:"title"`
	Description *string    `json:"description,omitempty"`
	Priority    *int       `json:"priority,omitempty"`
	State       string     `json:"state"`
	BranchName  *string    `json:"branch_name,omitempty"`
	URL         *string    `json:"url,omitempty"`
	Labels      []string   `json:"labels,omitempty"`
	BlockedBy   []Blocker  `json:"blocked_by,omitempty"`
	CreatedAt   *time.Time `json:"created_at,omitempty"`
	UpdatedAt   *time.Time `json:"updated_at,omitempty"`
}

// Blocker references another issue that gates dispatch of this one.
//
// SPEC §16.2: a Todo issue with any non-terminal blocker is held back.
type Blocker struct {
	ID         string `json:"id"`
	Identifier string `json:"identifier"`
	State      string `json:"state"`
}

// NormalizedState returns the lowercased state per SPEC §4.2.
func (i *Issue) NormalizedState() string {
	return strings.ToLower(i.State)
}

// BlockersAllTerminal reports whether every entry in BlockedBy is in one
// of the configured terminal states (case-insensitive).
//
// An issue with no blockers trivially satisfies this predicate.
func (i *Issue) BlockersAllTerminal(terminalStates []string) bool {
	if len(i.BlockedBy) == 0 {
		return true
	}
	for _, b := range i.BlockedBy {
		matched := false
		for _, ts := range terminalStates {
			if strings.EqualFold(b.State, ts) {
				matched = true
				break
			}
		}
		if !matched {
			return false
		}
	}
	return true
}
